lib/charms/synapse/v0/matrix_auth.py

```diff
@@ -0,0 +1,516 @@
+        self.framework.observe(self.matrix_auth.on.matrix_auth_request_processed, self._handler)
+        ...
+
+    def _handler(self, events: MatrixAuthRequestProcessed) -> None:
+        ...
+
+```
+
+As shown above, the library provides a custom event to handle the scenario in
+which a matrix authentication (homeserver and shared secret) has been added or updated.
+
+The MatrixAuthRequires provides an `update_relation_data` method to update the relation data by
+passing a `MatrixAuthRequirerData` data object, requesting a new authentication.
+
+### Provider Charm
+
+Following the previous example, this is an example of the provider charm.
+
+```python
+from charms.synapse.v0.matrix_auth import MatrixAuthProvides
+
+class MatrixAuthProviderCharm(ops.CharmBase):
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.plugin_auth = MatrixAuthProvides(self)
+        ...
+
+```
+The MatrixAuthProvides object wraps the list of relations into a `relations` property
+and provides an `update_relation_data` method to update the relation data by passing
+a `MatrixAuthRelationData` data object.
+
+```python
+class MatrixAuthProviderCharm(ops.CharmBase):
+    ...
+
+    def _on_config_changed(self, _) -> None:
+        for relation in self.model.relations[self.plugin_auth.relation_name]:
+            self.plugin_auth.update_relation_data(relation, self._get_matrix_auth_data())
+
+```
+"""
+
+# The unique Charmhub library identifier, never change it
+LIBID = "ff6788c89b204448b3b62ba6f93e2768"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+# pylint: disable=wrong-import-position
+import json
+import logging
+from typing import Dict, List, Optional, Tuple, cast
+
+import ops
+from pydantic import (
+    BaseModel,
+    Field,
+    SecretStr,
+)
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_RELATION_NAME = "matrix-plugin-auth"
+
+
+#### Data models for Provider and Requirer ####
+class MatrixAuthProviderData(BaseModel):
+    """Represent the MatrixAuth provider data.
+
+    Attributes:
+        homeserver: the homeserver URL.
+        shared_secret: the Matrix shared secret.
+        shared_secret_id: the shared secret Juju secret ID.
+    """
+
+    homeserver: Optional[str] = None
+    shared_secret: Optional[SecretStr] = Field(default=None, exclude=True)
+    shared_secret_id: Optional[SecretStr] = Field(default=None)
+
+    def set_shared_secret_id(self, model: ops.Model, relation: ops.Relation) -> None:
+        """Store the Matrix shared secret as a Juju secret.
+
+        Args:
+            model: the Juju model
+            relation: relation to grant access to the secrets to.
+        """
+        # password is always defined since pydantic guarantees it
+        password = cast(SecretStr, self.shared_secret)
```

Comment:
no action comment: it's unfortunate to have so many different things associated with "secret" word. Juju secret, Matrix shared secret, SecretStr.